### DISCUSSION TITLE
How do I use an async function?

### DISCUSSION THREAD

**Discussion by Xalwer** [DisnakeDev/disnake#1153] [Q&A] [answered]
```
# vc is disnake.VoiceClient

def after(error):
    await vc.disconnect()

vc.play(after=after)
```
How do I use `await vc.disconnect()`?

**Comment by ironman9356** [comment]
Hi, 
First you should know what `async` and `await` are. 
In simple terms: 

"async and await make promises easier to write"

async makes a function return a Promise

await makes a function wait for a Promise

[from here](https://www.w3schools.com/Js/js_async.asp)

So whenever you are using an `await` you must make sure its an `async` function 
In this case your `after` function must be `async` because its making a "promise" to do things in the function and the `await` inside the function returns the "promise" 

Example:
You are making a "promise" to the program that your bot will disconnect after some time. So you make an `async` function and make the function return the "promise" of bot disconnect. The `await` will return the "promise".
I hope this helps

**Reply by Xalwer** [reply]
the problem is I can't pass the async function as an argument to `vc.play(after=)`

this code will throw an error:
```
# vc is disnake.VoiceClient

async def after(error):
    await vc.disconnect()

vc.play(after=await after)
```
In general, I need the bot to simply disconnect after the end of the audio track.

Thanks for the answer.

**Reply by Sharp-Eyes** [reply, marked as answer]
I personally find something like this a fairly elegant solution:
```py
event: asyncio.Future[None] = asyncio.Future()

def after(exc: Exception | None):
    event.set_exception(exc) if exc else event.set_result(None)

vc.play(src, after=after)
await event  # wait until the player is done

await vc.disconnect()
```
The trick is to use something you can synchronously set a flag on and asynchronously wait for that flag to be set, which is something `asyncio.Future` is particularly well-suited for.
